cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlMetadataMigrationService.java

```diff
@@ -0,0 +1,121 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.sourcecontrol;
+
+import com.google.common.util.concurrent.AbstractScheduledService;
+import com.google.inject.Inject;
+import io.cdap.cdap.app.store.Store;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.internal.app.store.AppMetadataStore;
+import io.cdap.cdap.internal.app.store.NamespaceSourceControlMetadataStore;
+import io.cdap.cdap.proto.id.ApplicationReference;
+import io.cdap.cdap.proto.sourcecontrol.SourceControlMeta;
+import io.cdap.cdap.spi.data.transaction.TransactionRunner;
+import io.cdap.cdap.spi.data.transaction.TransactionRunners;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.apache.twill.common.Threads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Service for migrating source control metadata from app_spec table to source control table.
+ * Starting from 6.10.1 we have introduced a separate table for storing scm metadata. As this
+ * information was previously stored in app_spec table we need to run this migration
+ */
+public class SourceControlMetadataMigrationService extends AbstractScheduledService {
+
+  private ScheduledExecutorService executor;
+  private final long runInterval;
+  private final TransactionRunner transactionRunner;
+  private final Store store;
+
+  private static final Logger LOG = LoggerFactory.getLogger(SourceControlMetadataMigrationService.class);
+
+  @Inject
+  SourceControlMetadataMigrationService(CConfiguration cConf, Store store,
+      TransactionRunner transactionRunner) {
+    this.runInterval = cConf.getLong(
+        Constants.SourceControlManagement.METADATA_MIGRATION_INTERVAL_SECONDS);
+    this.transactionRunner = transactionRunner;
+    this.store = store;
+  }
+
+  @Override
+  protected final ScheduledExecutorService executor() {
+    executor = Executors.newSingleThreadScheduledExecutor(
+        Threads.createDaemonThreadFactory("scm-metadata-migration"));
+    return executor;
+  }
+
+  @Override
+  protected void runOneIteration() {
+    try {
+      migrateMetadata();
+      // stop if the migration is successfull otherwise retry after delay
+      this.stop();
+    } catch (Exception e) {
+      // Log and retry on exception
+      LOG.error("Failed to migrate source control metadata", e);
+    }
+  }
+
+  @Override
+  protected Scheduler scheduler() {
+    return Scheduler.newFixedRateSchedule(1, runInterval, TimeUnit.SECONDS);
```

Comment:
I'd recommend newFixedDelaySchedule to ensure there is always delay between runs. Also you can put 0 as first run delay.

Reply:
Done